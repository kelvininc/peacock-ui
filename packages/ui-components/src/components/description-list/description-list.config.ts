import { ComputePositionConfig, offset } from '@floating-ui/dom';
import { ETooltipPosition, IDescriptionListItemToggletipConfig } from '../../types';

const TEXT_TOOLTIP_MAIN_AXIS_OFFSET = 4;

export const DEFAULT_TEXT_TOOLTIP_CONFIG: Partial<ComputePositionConfig> = {
	placement: ETooltipPosition.Bottom,
	middleware: [offset(TEXT_TOOLTIP_MAIN_AXIS_OFFSET)]
};

export const DEFAULT_ICON_TOGGLETIP_CONFIG: IDescriptionListItemToggletipConfig = {
	position: ETooltipPosition.BottomStart
};
