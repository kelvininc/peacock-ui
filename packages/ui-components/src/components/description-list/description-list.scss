@import '../../assets/styles/globals';

.description-list-container {
	display: grid;
	grid-template-columns: max-content 1fr;
	gap: $spacing-5x $spacing-7x;
	word-break: break-all;
	align-items: baseline;

	.title {
		@include kv-font-label-small-uppercase-regular;
		color: kv-color('neutral-4');
	}

	.description {
		@include kv-font-paragraph-regular;
		color: kv-color('neutral-2');

		kv-tooltip {
			&::part(content) {
				display: inline;
			}

			&::part(container) {
				padding: $spacing-2x;
			}
		}

		kv-toggle-tip {
			&::part(toggle-tip-open-element-container) {
				display: inline-block;
				vertical-align: text-top;
				margin-left: $spacing-2x;
			}

			&::part(toggle-tip-container) {
				padding: $spacing-2x;
			}
		}

		kv-icon::part(icon) {
			cursor: pointer;
		}
	}
}
