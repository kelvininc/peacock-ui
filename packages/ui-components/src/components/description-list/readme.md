# _<kv-description-list>_



<!-- Auto Generated Below -->


## Usage

### Angular

```html

<!-- Default -->
const defaultConfig: IDescriptionListItem[] = [
	{
		title: 'Name ID',
		description: 'cluster-a-brownie',
		popoverInfo: {
			text: `This name cannot be changed`,
			icon: EIconName.Info
		}
	},
	{
		title: 'Kubernetes Version',
		description: '1.20.5'
	},
	{
		title: 'Kelvin Version',
		description: '4.2.4'
	}
];
<kv-description-list items=[defaultConfig] />

<!-- With text tooltip -->
const withTextTooltip: IDescriptionListItem[] = [
	{
		title: 'Name ID',
		description: 'cluster-a-brownie',
		popoverInfo: {
			text: `This name cannot be changed`
		}
	},
	{
		title: 'Kubernetes Version',
		description: '1.20.5'
	},
	{
		title: 'Kelvin Version',
		description: '4.2.4'
	}
];
<kv-description-list items=[withTextTooltip] />

<!-- With icon tooltip -->
const withIconTooltip: IDescriptionListItem[] = [
	{
		title: 'Name ID',
		description: 'cluster-a-brownie',
		popoverInfo: {
			text: `This name cannot be changed`,
			icon: EIconName.Info
		}
	},
	{
		title: 'Kubernetes Version',
		description: '1.20.5'
	},
	{
		title: 'Kelvin Version',
		description: '4.2.4'
	}
];
<kv-description-list items=[withIconTooltip] />
```


### React

```tsx
import React from 'react';

import { KvDescriptionList } from '@kelvininc/react-ui-components';

const defaultConfig: IDescriptionListItem[] = [
	{
		title: 'Name ID',
		description: 'cluster-a-brownie',
		popoverInfo: {
			text: `This name cannot be changed`,
			icon: EIconName.Info
		}
	},
	{
		title: 'Kubernetes Version',
		description: '1.20.5'
	},
	{
		title: 'Kelvin Version',
		description: '4.2.4'
	}
];

const withTextTooltip: IDescriptionListItem[] = [
	{
		title: 'Name ID',
		description: 'cluster-a-brownie',
		popoverInfo: {
			text: `This name cannot be changed`
		}
	},
	{
		title: 'Kubernetes Version',
		description: '1.20.5'
	},
	{
		title: 'Kelvin Version',
		description: '4.2.4'
	}
];

const withIconTooltip: IDescriptionListItem[] = [
	{
		title: 'Name ID',
		description: 'cluster-a-brownie',
		popoverInfo: {
			text: `This name cannot be changed`,
			icon: EIconName.Info
		}
	},
	{
		title: 'Kubernetes Version',
		description: '1.20.5'
	},
	{
		title: 'Kelvin Version',
		description: '4.2.4'
	}
];

export const KvDescriptionListExample: React.FC = () => (
	<>
		{/*-- Default --*/}
		<KvDescriptionList items={defaultConfig} />

		{/*-- With text tooltip --*/}
		<KvDescriptionList items={withTextTooltip} />

		{/*--  With icon tooltip --*/}
		<KvDescriptionList items={withIconTooltip} />
	</>
);
```



## Properties

| Property                   | Attribute | Description                                                              | Type                                                                                                                                                                                                                                                                                                                                                                           | Default                         |
| -------------------------- | --------- | ------------------------------------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | ------------------------------- |
| `descriptionTooltipConfig` | --        | (optional) The config to use on the tooltip that shows hovering the text | `{ placement?: Placement; strategy?: Strategy; middleware?: (false \| { name: string; options?: any; fn: (state: { platform: Platform; placement: Placement; strategy: Strategy; x: number; y: number; initialPlacement: Placement; middlewareData: MiddlewareData; rects: ElementRects; elements: Elements; }) => Promisable<MiddlewareReturn>; })[]; platform?: Platform; }` | `DEFAULT_TEXT_TOOLTIP_CONFIG`   |
| `iconToggletipConfig`      | --        | (optional) The config to use on the icon toggletip                       | `IDescriptionListItemToggletipConfig`                                                                                                                                                                                                                                                                                                                                          | `DEFAULT_ICON_TOGGLETIP_CONFIG` |
| `items` _(required)_       | --        | (required) The array of items to display in the list                     | `IDescriptionListItem[]`                                                                                                                                                                                                                                                                                                                                                       | `undefined`                     |


## Dependencies

### Depends on

- [kv-tooltip](../tooltip)
- [kv-toggle-tip](../toggle-tip)
- [kv-icon](../icon)

### Graph
```mermaid
graph TD;
  kv-description-list --> kv-tooltip
  kv-description-list --> kv-toggle-tip
  kv-description-list --> kv-icon
  style kv-description-list fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


