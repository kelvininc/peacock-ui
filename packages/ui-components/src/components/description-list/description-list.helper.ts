import { IDescriptionListItemPopover } from './description-list.types';

export function getTooltipText(popoverInfo: IDescriptionListItemPopover) {
	return popoverInfo?.text && !popoverInfo.icon ? popoverInfo.text : '';
}
