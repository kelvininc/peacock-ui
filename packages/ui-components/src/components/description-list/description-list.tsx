import { Component, Fragment, Host, Prop, h } from '@stencil/core';
import { IDescriptionList, IDescriptionListItem, IDescriptionListItemToggletipConfig } from './description-list.types';
import { DEFAULT_ICON_TOGGLETIP_CONFIG, DEFAULT_TEXT_TOOLTIP_CONFIG } from './description-list.config';
import { getTooltipText } from './description-list.helper';
import { ComputePositionConfig } from '@floating-ui/dom';

@Component({
	tag: 'kv-description-list',
	styleUrl: 'description-list.scss',
	shadow: true
})
export class KvDescriptionList implements IDescriptionList {
	/** @inheritdoc */
	@Prop({ reflect: true }) items!: IDescriptionListItem[];
	/** @inheritdoc */
	@Prop({ reflect: true }) descriptionTooltipConfig?: Partial<ComputePositionConfig> = DEFAULT_TEXT_TOOLTIP_CONFIG;
	/** @inheritdoc */
	@Prop({ reflect: true }) iconToggletipConfig?: IDescriptionListItemToggletipConfig = DEFAULT_ICON_TOGGLETIP_CONFIG;

	render() {
		return (
			<Host>
				<div class="description-list-container">
					{this.items?.map(({ title, description, popoverInfo }) => (
						<Fragment>
							<div class="title">{title}</div>
							<div class="description">
								<kv-tooltip text={getTooltipText(popoverInfo)} options={this.descriptionTooltipConfig}>
									{description}
								</kv-tooltip>
								{popoverInfo?.icon && (
									<kv-toggle-tip text={popoverInfo.text} {...this.iconToggletipConfig} exportparts="toggle-tip-container">
										<kv-icon name={popoverInfo.icon} customClass={'icon-16'} slot="open-element-slot" />
									</kv-toggle-tip>
								)}
							</div>
						</Fragment>
					))}
				</div>
			</Host>
		);
	}
}
