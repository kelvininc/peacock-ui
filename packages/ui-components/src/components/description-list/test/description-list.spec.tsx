import { h } from '@stencil/core';
import { SpecPage, newSpecPage } from '@stencil/core/testing';
import { KvDescriptionList } from '../description-list';
import { DEFAULT_CONFIG, WITH_ICON_TOOLTIP, WITH_TEXT_TOOLTIP } from './description-list.mock';

describe('Description List (unit tests)', () => {
	let page: SpecPage;

	describe('when rendering with required item props', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvDescriptionList],
				template: () => <kv-description-list items={DEFAULT_CONFIG} />
			});
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when rendering with an item with text tooltip', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvDescriptionList],
				template: () => <kv-description-list items={WITH_TEXT_TOOLTIP} />
			});
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when rendering with an item with icon tooltip', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvDescriptionList],
				template: () => <kv-description-list items={WITH_ICON_TOOLTIP} />
			});
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});
});
