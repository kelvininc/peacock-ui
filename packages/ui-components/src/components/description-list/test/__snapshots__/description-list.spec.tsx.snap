// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Description List (unit tests) when rendering with an item with icon tooltip should match the snapshot 1`] = `
<kv-description-list>
  <mock:shadow-root>
    <div class="description-list-container">
      <div class="title">
        Name ID
      </div>
      <div class="description">
        <kv-tooltip text="">
          cluster-a-brownie
        </kv-tooltip>
        <kv-toggle-tip exportparts="toggle-tip-container" position="bottom-start" text="This name cannot be changed">
          <kv-icon customclass="icon-16" name="kv-info" slot="open-element-slot"></kv-icon>
        </kv-toggle-tip>
      </div>
      <div class="title">
        Kubernetes Version
      </div>
      <div class="description">
        <kv-tooltip text="">
          1.20.5
        </kv-tooltip>
      </div>
      <div class="title">
        Kelvin Version
      </div>
      <div class="description">
        <kv-tooltip text="">
          4.2.4
        </kv-tooltip>
      </div>
    </div>
  </mock:shadow-root>
</kv-description-list>
`;

exports[`Description List (unit tests) when rendering with an item with text tooltip should match the snapshot 1`] = `
<kv-description-list>
  <mock:shadow-root>
    <div class="description-list-container">
      <div class="title">
        Name ID
      </div>
      <div class="description">
        <kv-tooltip text="This name cannot be changed">
          cluster-a-brownie
        </kv-tooltip>
      </div>
      <div class="title">
        Kubernetes Version
      </div>
      <div class="description">
        <kv-tooltip text="">
          1.20.5
        </kv-tooltip>
      </div>
      <div class="title">
        Kelvin Version
      </div>
      <div class="description">
        <kv-tooltip text="">
          4.2.4
        </kv-tooltip>
      </div>
    </div>
  </mock:shadow-root>
</kv-description-list>
`;

exports[`Description List (unit tests) when rendering with required item props should match the snapshot 1`] = `
<kv-description-list>
  <mock:shadow-root>
    <div class="description-list-container">
      <div class="title">
        Name ID
      </div>
      <div class="description">
        <kv-tooltip text="">
          cluster-a-brownie
        </kv-tooltip>
      </div>
      <div class="title">
        Kubernetes Version
      </div>
      <div class="description">
        <kv-tooltip text="">
          1.20.5
        </kv-tooltip>
      </div>
      <div class="title">
        Kelvin Version
      </div>
      <div class="description">
        <kv-tooltip text="">
          4.2.4
        </kv-tooltip>
      </div>
    </div>
  </mock:shadow-root>
</kv-description-list>
`;
